### PR TITLE
cdist meta function

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79051
* #79042
* #79026
* #78997
* #78996
* __->__ #78993
* #78992

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -311,3 +311,22 @@ def meta_addbmm(self, batch1, batch2, *, beta=1, alpha=1):
 
 torch.library.impl(meta_lib, "addbmm")(meta_addbmm)
 torch.library.impl(meta_lib, "addbmm.out")(meta_addbmm)
+
+@torch.library.impl(meta_lib, "_cdist_forward")
+def meta_cdist_forward(x1, x2, p, compute_mode):
```
Comment (Lezcano) [inline torch/_meta_registrations.py:316]: Missing the `compute_mode` checks (it's an int between 0 and 2).

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -311,3 +311,22 @@ def meta_addbmm(self, batch1, batch2, *, beta=1, alpha=1):
...
+    check(utils.is_float_dtype(x1.dtype), lambda: "cdist only supports floating-point dtypes, X1 got: {x1.dtype}")
+    check(utils.is_float_dtype(x2.dtype), lambda: "cdist only supports floating-point dtypes, X2 got: {x2.dtype}")
+    check(p >= 0, lambda: "cdist only supports non-negative p values")
+    r1 = x1.size(-2)
+    r2 = x2.size(-2)
+    batch_tensor1 = x1.shape[:-2]
+    batch_tensor2 = x2.shape[:-2]
+    output_shape = list(torch.broadcast_shapes(batch_tensor1, batch_tensor2))
+    output_shape.extend([r1, r2])
+    return x1.new_empty(output_shape)
```
Comment (Lezcano) [inline torch/_meta_registrations.py:332]: The logic in the original function is more complicated. Are the two equivalent? It'd make sense to me that this logic in the meta functiion is all it's going on in the original function, but I have not checked.

Comment (ezyang) [reply torch/_meta_registrations.py:333]: This one I worked out by just reading the docs. If the implementation disagrees it's probably a bug

**Review by lezcano — APPROVED** [review]

File: torch/_meta_registrations.py (hunk 2 above)
Comment (Lezcano) [reply torch/_meta_registrations.py:333]: The implementation looks a bit messy. It may very well just do this, but it's not clear at first sight.

**Review by Chillee — APPROVED** [review]

**Pull request closed without merging**